src/jdk.compiler/share/classes/com/sun/tools/javac/comp/TypeEnter.java

```diff
@@ -332,6 +334,26 @@ private void importJavaLang(JCCompilationUnit tree, Env<AttrContext> env, Import
             }
             importAll(make.at(tree.pos()).Import(make.Select(make.QualIdent(javaLang.owner), javaLang), false),
                 javaLang, env);
+
+            List<JCTree> defs = tree.defs;
+            boolean isImplicitClass = !defs.isEmpty() &&
+                    defs.head instanceof JCClassDecl cls &&
+                    (cls.mods.flags & IMPLICIT_CLASS) != 0;
+            if (isImplicitClass) {
+                doModuleImport(make.ModuleImport(make.QualIdent(syms.java_base)));
+                if (peekTypeExists(env, syms.ioType.tsym)) {
+                    doImport(make.Import(make.Select(make.QualIdent(syms.ioType.tsym),
+                            names.asterisk), true));
+                }
+            }
+        }
+
+        //check if the given class exists, without producing side-effects:
+        private boolean peekTypeExists(Env<AttrContext> env, TypeSymbol type) {
+            JCExpression expr = make.Select(make.QualIdent(type), names._class);
+            return !deferredAttr.attribSpeculative(expr, env, attr.unknownExprInfo)
```

Comment:
Is this code meant to be transient?

Note that we have an handy method in Symtab called `synthesizeEmptyInterfaceIfMissing` which will lookup a type, then, if such a type is missing, it will replace it with an empty interface. Since IO is a class, perhaps we can just check if the symbol in the symbol table is an interface or not, and avoid deferred attribution?

Reply:
It was meant to be permanent, even though now transiently it is more important. Long term, it is much less important for the implicit imports in implicit classes. The String Templates code was trying to detect if the `StringTemplate` class exists and avoid the implicit import if it didn't. That didn't work very well, and was causing issues in some corner cases. So, I used the speculative attribution, and used it for implicitly declared classes as well.

The reason why I used the speculative attribution is that it does not have side-effects - i.e. if the class does not exist, and the user types `java.io.IO` explicitly, there will still be an error. With `synthetizeEmptyInterfaceIfMissing`, there will be no error reported, even for explicit uses of the type, I believe.

Reply:
Maybe we should also set `SYNTHETIC` in `synthetizeEmptyInterfaceIfMissing` ? That method was mostly added so that a "real" type was available to javac. That said, if you didn't want to mess with that I understand, but the general approach seem to remain valid - e.g. in principle we should be able to wrap the `IO`'s symbol completer, so that if a completion error occurs, the error is swallowed (and then we know the type doesn't exist). In other words, a deferred attribution seems an heavy way to see if a type is there?

Reply:
I still think the `synthetizeEmptyInterfaceIfMissing` is a bit too optimistic - that makes the type "available" to javac even if does not exist, and that is probably not correct in this case. (It would mean one could write `java.io.IO io;`, and not get an error even if type didn't exist.)

But, I admit the speculative attribution was probably a bit too pessimistic. I've changed the code to do attribution, and check the result. Does this look better?

Thanks!

Reply:
> I still think the `synthetizeEmptyInterfaceIfMissing` is a bit too optimistic - that makes the type "available" to javac even if does not exist, and that is probably not correct in this case. (It would mean one could write `java.io.IO io;`, and not get an error even if type didn't exist.)
> 
> But, I admit the speculative attribution was probably a bit too pessimistic. I've changed the code to do attribution, and check the result. Does this look better?
> 
> Thanks!

Thanks for the changes, the new approach seems a lot more balanced.